libbeat/common/reload/reload.go

```diff
@@ -0,0 +1,105 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package reload
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+var register = newRegistry()
+
+// ConfigWithMeta holds a pair of common.Config and optional metadata for it
+type ConfigWithMeta struct {
+	// Config to store
+	Config *common.Config
+
+	// Meta data related to this config
+	Meta *common.MapStrPointer
+}
+
+// ReloadableList provides a method to reload the configuration of a list of entities
+type ReloadableList interface {
+	Reload(configs []*ConfigWithMeta) error
+}
+
+// Reloadable provides a method to reload the configuration of an entity
+type Reloadable interface {
+	Reload(config *ConfigWithMeta) error
+}
+
+type registry struct {
+	sync.RWMutex
+	confsLists map[string]ReloadableList
+	confs      map[string]Reloadable
+}
+
+func newRegistry() registry {
+	return registry{
+		confsLists: make(map[string]ReloadableList),
+		confs:      make(map[string]Reloadable),
+	}
+}
+
+// MustRegister declares a reloadable object
+func MustRegister(name string, r Reloadable) {
+	register.Lock()
+	defer register.Unlock()
+
+	if r == nil {
+		panic("Got a nil object")
+	}
+
+	if _, ok := register.confs[name]; ok {
+		panic(fmt.Sprintf("%s configuration is already registered", name))
+	}
+
+	register.confs[name] = r
+}
+
+// MustRegisterList declares a reloadable list of configurations
+func MustRegisterList(name string, list ReloadableList) {
+	register.Lock()
+	defer register.Unlock()
+
+	if list == nil {
+		panic("Got a nil object")
+	}
+
+	if _, ok := register.confsLists[name]; ok {
+		panic(fmt.Sprintf("%s configuration list is already registered", name))
+	}
+
+	register.confsLists[name] = list
+}
+
+// Get returns the reloadable object with the given name, nil if not found
+func Get(name string) Reloadable {
+	register.RLock()
+	defer register.RUnlock()
+	return register.confs[name]
+}
+
+// GetList returns the reloadable list with the given name, nil if not found
+func GetList(name string) ReloadableList {
+	register.RLock()
+	defer register.RUnlock()
+	return register.confsLists[name]
+}
```

Comment:
That's an option, but in practice, that would mean we need to pass an instance of the registry to all the running parts of the beat. Is there any plan to introduce something like `pkg/context` for this?

Reply:
@exekias we want to introduce this, but we are a bit far from that, I am toying with the idea in beatless.

Let me clarify, when I first read your PR, I thought why not use the global feature registry to hold that and just wrap it in some function calls to do the type assertions, this is what I am doing in the [beatless PR](https://github.com/elastic/beats/pull/8170).

But after taking a step back I thought it's probably better to have this specific registry which deal with alive config changes. But I am not sure how you want to structure it with the application but I would expect some kind of listener that will trigger events? Maybe a bit of details or pointers of the usage would help me?

With that in mind I think what we can do without introducing a `pkg/context` is to make sure that the registry you create is instead a type that your Registry variable instantiate, this is what you almost do.

But instead of exposing the internals of the registry in the exposed function like the following you do like the next example which hide the implementation details but provide package accessable function. Will help if you add more logic to the class to be able to unit test it without having side effects on the running code.

```golang

func MustRegister(name string, r Reloadable) {
	register.Lock()
	defer register.Unlock()
	if r == nil {
		panic("Got a nil object")
	}
	if _, ok := register.confs[name]; ok {
		panic(fmt.Sprintf("%s configuration is already registered", name))
	}
	register.confs[name] = r
}

var Registry ...

func MustRegister(name string, r Reloadable) {
    err := Registry.Register(name, r)
    if err != nil {
     panic(".... something bad happened report the error.")
    }
}



```

Reply:
I think I got it, will do the changes! Thanks for your input :+1: